### PR TITLE
Use dotenv to conditionally apply .env file

### PR DESCRIPTION
**Pull request by sareh**
Part of [this Simorgh Infra ticket](https://github.com/bbc/simorgh-infrastructure/issues/194). Part of (possibly all?) #1158.

Quite a lot of moving parts here, so here goes. This PR:

- Removes all examples of `export $(cat env)...`, in favour of adding `dotenv` dependency, which will handle this for us. When called in code, `dotenv` will establish `process.env.___` variables with values from the `.env` file, **unless there is an existing environment variable with a different value**.
- Add `spec` section to `package.json`. This lines up with [some changes in Simorgh Infrastructure](https://github.com/bbc/simorgh-infrastructure/pull/192). This section now ensures the Simorgh RPM will make bake-scripts executable & move them to the correct directory post-RPM-install.

Original, outdated description:

---

**Background**
With the current use of `.env` file, we cannot override the environment variables with a `SIMORGH_BASE_URL=https://www.test.bbc.com` prior to a command such as  `npm run build` on CI.

The only way of overwriting will be to actively overwrite the `.env` file within the Simorgh directory from the Jenkinsfile in Simorgh Infrastructure, as we build our application on TEST. I think that this is not an ideal way of doing things, since we will be changing files within our application directory from our Infrastructure build and this is a bad precedent to set. 

I propose that we remove this local `.env` file and move to having local environment variables defined by running a script when setting up your local developer environment for the first time `npm run setupLocalEnvVariables`.

This will mean that overriding environment variables in both Jenkins and for Cypress tests should be very simple:
e.g. for a cypress test we could build the application with Test base path & check that the assets have the right paths:
`SIMORGH_BASE_URL=https://www.test.bbc.com npm run build` etc.

For the Test & Live Jenkins environments we can update these files to have the relevant key/value pairs `SIMORGH_BASE_URL`, etc.` https://github.com/bbc/simorgh-infrastructure/blob/latest/infrastructure/config/test-simorgh-cosmosconfig.json

- [x] I have assigned myself to this PR and the corresponding issues
- [ ] Tests added for new features
- [ ] Test engineer approval
- [ ] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

### CODE REVIEW

**Review by ChrisBAshton — COMMENTED** [review]
Thanks for explaining, the reasoning makes sense. My only question is the implementation - it's not ideal to have to run a `setupLocalEnvVariables` script every time we turn on our machines or open a new terminal tab.

I read that [dotenv doesn't modify pre-existing ENV variables](https://www.npmjs.com/package/dotenv#what-happens-to-environment-variables-that-were-already-set), so we could try using that. I had difficulties using it before, but didn't realise it ignored pre-existing ENV variables, so it may have been to do with that!

That way we would just `require('dotenv')` at the top of our webpack.config.js (and Jest) and should be ok!

**Comment by sareh** [comment]
@ChrisBAshton I'd be happy to move to `dotenv` since it would be a much cleaner implementation and get the same results.

**Comment by bcmn** [comment]
> I read that dotenv doesn't modify pre-existing ENV variables, so we could try using that. I had difficulties using it before, but didn't realise it ignored pre-existing ENV variables, so it may have been to do with that!

Struggling to recreate this behaviour. Running both ` PUBLIC_PATH=___ RAZZLE_BASE_PATH=___ SIMORGH_BASE_URL=___ npm run build` & `export PUBLIC_PATH=___ && export RAZZLE_BASE_PATH=___ && export SIMORGH_BASE_URL=___ && npm run build` still result in the `.env` values being assigned.

**Comment by bcmn** [comment]
Hm. So I have observed the behaviour above when creating a node script to test it. So I think something Simorgh is doing must be overriding it? 🤔

**Comment by bcmn** [comment]
screenshot 2019-01-14 at 15 15 40

[Have got this behaviour working on a branch](https://ci.news.tools.bbc.co.uk/job/simorgh/job/remove-env-file-test/2/consoleFull), in combination with Phil's suggested change around the Local getInitialData URL. Gonna try deploying to TEST once more & see how things go.

File: src/server/index.jsx [hunk 1]
```diff
@@ -11,12 +11,19 @@ import path from 'path';
...
 import helmet from 'helmet';
 import gnuTP from 'gnu-terry-pratchett';
+import dotenv from 'dotenv';
 import routes, { articleRegexPath } from '../app/routes';
 import { getStyleTag } from './styles';
 import getAssetsArray from './assets';
 
 import Document from '../app/components/Document';
 
+const result = dotenv.config();
```
Comment (dr3) [inline src/server/index.jsx:21]: While we do need the env variables here, we also need them in webpack, can we also do this in the various webpack files?

**Review by dr3 — CHANGES_REQUESTED** [review]

File: src/server/index.jsx (hunk 1 above)
Comment (bcmn) [reply src/server/index.jsx:21]: Good catch -- done.

Comment (ChrisBAshton) [reply src/server/index.jsx:21]: Should only need it in the `webpack.config.js` file - the other webpack config files are pulled into that one.

File: webpack.config.client.js [hunk 2]
```diff
@@ -1,6 +1,13 @@
 /* eslint-disable global-require */
 const AssetsPlugin = require('assets-webpack-plugin');
 const CopyWebpackPlugin = require('copy-webpack-plugin');
+const dotenv = require('dotenv');
+
+const result = dotenv.config();
+
+if (result.error) {
+  throw result.error;
+}
```
Comment (dr3) [inline webpack.config.client.js]: This is, somehow, causing files to be requested from `http://localhost:7080/news/articles/undefined/static/js/vendor-678f84af.a7416d5c.js` locally on build & start

File: src/app/routes/getInitialData/index.js [hunk 3]
```diff
@@ -1,20 +1,12 @@
...
-};
-
 const getInitialData = async ({ match }) => {
   try {
     const { id, service, amp } = match.params;
 
-    const url = `${getBaseUrl()}/${service}/articles/${id}.json`;
+    const url = `${
+      process.env.SIMORGH_BASE_URL
+    }/${service}/articles/${id}.json`;
```
Comment (dr3) [inline src/app/routes/getInitialData/index.js]: My browser locally is requesting http://localhost:7080/news/articles/undefined/news/articles/c85pqyj5m2ko.json here

Comment (dr3) [reply src/app/routes/getInitialData/index.js]: The client doesnt have any idea of process, I think we need to take this https://github.com/bbc/simorgh/pull/871/files#diff-00bc7ce2bc299041871f59a72a2eb568R12

Comment (bcmn) [reply src/app/routes/getInitialData/index.js]: Thought this over on the way home. Is this a separate issue? Fix the getting environment variables stuff, revert the changes in `src/app/routes/getInitialData/index.js`, then tackle Phil's issue in a separate PR which addresses the fact we no longer have build-time variables from Razzle? Would stop this PR becoming too big.

**Comment by bcmn** [comment]
To clarify: this ticket no longer applies the change described in #1168. This will be done in a follow-up PR. I have also opened #1181.

**Review by dr3 — DISMISSED** [review]
👍 LGTM, Will only half work untill the webpack process envs are fixed

**Review by pjlee11 — DISMISSED** [review]
GREAT work!

File: webpack.config.client.js [hunk 4]
```diff
@@ -1,6 +1,13 @@
 /* eslint-disable global-require */
 const AssetsPlugin = require('assets-webpack-plugin');
 const CopyWebpackPlugin = require('copy-webpack-plugin');
+const dotenv = require('dotenv');
```
Comment (ChrisBAshton) [inline webpack.config.client.js]: Should this not be in the base `webpack.config.js` file? That way it's made available whichever of client/server we're compiling.

File: webpack.config.client.js [hunk 5]
```diff
@@ -1,6 +1,13 @@
...
+
+const result = dotenv.config();
```
Comment (ChrisBAshton) [inline webpack.config.client.js]: Can you add a comment explaining what we're doing here? Ideally any new starter can look at any part of our Webpack config and it'll be annotated for them.

**Review by ChrisBAshton — CHANGES_REQUESTED** [review]
> Add spec section to package.json. This lines up with some changes in Simorgh Infrastructure. This section now ensures the Simorgh RPM will make bake-scripts executable & move them to the correct directory post-RPM-install.

^ Is this strictly necessary for this PR or should the spec section be a separate PR? Not wanting to create work for you but the PR description has a _lot_ going on in it!

**Comment by bcmn** [comment]
> > Add spec section to package.json. This lines up with some changes in Simorgh Infrastructure. This section now ensures the Simorgh RPM will make bake-scripts executable & move them to the correct directory post-RPM-install.
> 
> ^ Is this strictly necessary for this PR or should the spec section be a separate PR? Not wanting to create work for you but the PR description has a _lot_ going on in it!

I believe so — that ensures the bake-scripts actually get run, which alters the `.env` file, which `dotenv` then imports.

**Review by ChrisBAshton — APPROVED** [review]
Awesome, thanks 👍

**Review by pjlee11 — APPROVED** [review]

**Comment by jamesbrumpton** [comment]
LGTM 👍

**Pull request merged**